src/DropDown/DropDown.tsx

```diff
@@ -1,71 +1,460 @@
-                {props.label && <label className="custom-label" htmlFor={props.name}>{props.label}</label>}
-
-                {props.native &&
-                    <select
-                        className={"dropdown-input form-control"}
-                        name={props.name}
-                        value={props.selectedValue ? props.selectedValue.value : ""}
-                        onChange={props.onChange}
-                    >
-                        {props.placeholder && <option key="default" value="DEFAULT">{props.placeholder}</option>}
-                        {props.list.map((item) =>
-                            <option key={item.value} value={item.value} >
-                                {item.label}
-                            </option>
-                        )}
-                    </select>
-                }
+    // COMPONENT INTERNAL STATE INIT ================================
+    const [open, setOpen] = React.useState(false);
+    const [shouldFocus, setShouldFocus] = React.useState(false);
+    const [currentFocused, setCurrentFocused] = React.useState(-1);
+    const [searchText, setSearchText] = React.useState("");
+
+    // REFS ================================
+    const dropdownToggleRef = React.createRef<HTMLDivElement>();
+    const dropdownMenuRef = React.createRef<HTMLDivElement>();
+    const searchRef = React.createRef<HTMLInputElement>();
+
+    // EFFECTS ================================
+    // Adding event listener to listen to clicks outside the component on mount, removing on unmount
+    React.useEffect(() => {
+        document.addEventListener("mousedown", handleClickOutside);
+        return () => {
+            document.removeEventListener("mousedown", handleClickOutside);
+        };
+    });
+
+    const handleClickOutside = (event) => {
+        if ((dropdownToggleRef.current && !dropdownToggleRef.current.contains(event.target)) && (dropdownMenuRef.current && !dropdownMenuRef.current.contains(event.target))) {
+            setOpen(false);
+        }
+    };
+
+    // As soon as the menu opens focus the search text input or the menu
+    React.useEffect(() => {
+        if (open) {
+            handleFocus();
+        } else {
+            setSearchText("");
+            if (currentFocused > -1) {
+                setCurrentFocused(-1);
+            }
+        }
+    }, [open]);
+
+    React.useEffect(() => {
+        if (open) {
+            handleFocus();
+        }
+    }, [currentFocused, shouldFocus]);
+
+    const handleFocus = (): void => {
+        const focusSuccess = focusCurrentItem();
+        if (!focusSuccess) {
+            setInitialFocus();
+        }
+    };
+
+    const focusCurrentItem = (): boolean => {
+        if (shouldFocus && listRefs[currentFocused] && listRefs[currentFocused].current) {
+            listRefs[currentFocused].current.focus();
+            return true;
+        }
+        return null;
```

Comment:
would it be fine to return null if this function type is boolean ?

Reply:
resolved in commit https://github.com/sebgroup/react-components/pull/26/commits/90ff6df057917ce882a2c165a3719737dfcc0a65